modules/codegen/src/main/scala/com/twilio/guardrail/generators/AkkaHttpServerGenerator.scala

```diff
@@ -522,88 +522,100 @@ object AkkaHttpServerGenerator {
-                case NonEmptyList(term, Nil)   => q"${term}.map(v1 => Right(Tuple1(v1)))"
-                case NonEmptyList(term, terms) => q"(..${term +: terms}).mapN(${Term.Name(s"Tuple${terms.length + 1}")}.apply).map(Right.apply)"
-              }}
+              case RouteMeta.UrlencodedFormData => {
+                val unmarshallerTerm = q"FormDataUnmarshaller"
+                val fru = q"""
+                  implicit val ${Pat.Var(unmarshallerTerm)}: FromRequestUnmarshaller[Either[Throwable, ${optionalTypes}]] =
+                    implicitly[FromRequestUnmarshaller[FormData]].flatMap { implicit executionContext => implicit mat => formData =>
+                      def unmarshalField[A: Decoder](name: String, value: String): Future[A] =
+                        Unmarshaller.firstOf(jsonDecoderUnmarshaller[A]).apply(value).recoverWith({
+                          case ex =>
+                            Future.failed(RejectionError(MalformedFormFieldRejection(name, ex.getMessage, Some(ex))))
+                        })
+
+                      ${params
+                  .map(
+                    param =>
+                      if (param.isFile) {
+                        q"Future.successful(Option.empty[(File, Option[String], ContentType)])"
+                      } else {
+                        val (realType, getFunc, transformResponse): (Type, Term.Name, (Term => Term)) = param.argType match {
+                          case t"Iterable[$x]"         => (x, q"getAll", (x: Term) => q"${x}.map(Option.apply)")
+                          case t"Option[Iterable[$x]]" => (x, q"getAll", (x: Term) => q"${x}.map(Option.apply)")
+                          case t"Option[$x]"           => (x, q"get", (x: Term) => x)
+                          case x                       => (x, q"get", (x: Term) => x)
+                        }
+                        transformResponse(
+                          q"""formData.fields.${getFunc}(${param.argName.toLit}).traverse(unmarshalField[${realType}](${param.argName.toLit}, _))"""
+                        )
+                    }
+                  ) match {
+                  case NonEmptyList(term, Nil)   => q"${term}.map(v1 => Right(Tuple1(v1)))"
+                  case NonEmptyList(term, terms) => q"(..${term +: terms}).mapN(${Term.Name(s"Tuple${terms.length + 1}")}.apply).map(Right.apply)"
+                }}
+                    }
+                """
+                (List(fru), NonEmptyList.one(unmarshallerTerm))
               }
-          """), Option(unmarshallerTerm))
-            } else (List.empty, None)
+
+              case RouteMeta.ApplicationJson => throw new Exception(s"Unable to generate unmarshaller for application/json")
+
+              case RouteMeta.OctetStream => throw new Exception(s"Unable to generate unmarshaller for application/octet-stream")
+
+              case RouteMeta.TextPlain => throw new Exception(s"Unable to generate unmarshaller for text/plain")
```

Comment:
Wartremover is gonna be pissed at you for these ;)

Reply:
:hide: